lib/chef/provider/file.rb

```diff
@@ -67,7 +67,13 @@ class File < Chef::Provider
       def initialize(new_resource, run_context)
         @content_class ||= Chef::Provider::File::Content
         if new_resource.respond_to?(:atomic_update)
-          @deployment_strategy = Chef::FileContentManagement::Deploy.strategy(new_resource.atomic_update)
+          # Resolves https://github.com/chef/chef/issues/4949
+          if docker_guest?(run_context.node) && (new_resource.path == "/etc/resolv.conf" ||
+                                                 new_resource.path == "/etc/hostname")
+            @deployment_strategy = Chef::FileContentManagement::Deploy.strategy(false)
+          else
+            @deployment_strategy = Chef::FileContentManagement::Deploy.strategy(new_resource.atomic_update)
```

Comment:
this could probably look more like:

```ruby
atomic_updates = special_docker_files? ? false : new_resource.atomic_updates
@deployment_strategy = Chef::FileContentManagement::Deploy.strategy(atomic_updates)
```

also should probably compare against `Pathname.cleanpath(new_resource.path)` and/or use `File.expand_path()`